libraries/fc/src/network/http/http_client.cpp

```diff
@@ -0,0 +1,361 @@
+#include <fc/network/http/http_client.hpp>
+#include <fc/io/json.hpp>
+#include <fc/scoped_exit.hpp>
+
+#include <boost/beast/core.hpp>
+#include <boost/beast/http.hpp>
+#include <boost/beast/version.hpp>
+#include <boost/asio/connect.hpp>
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/asio/ssl/error.hpp>
+#include <boost/asio/ssl/stream.hpp>
+
+using tcp = boost::asio::ip::tcp;       // from <boost/asio/ip/tcp.hpp>
+namespace http = boost::beast::http;    // from <boost/beast/http.hpp>
+namespace ssl = boost::asio::ssl;       // from <boost/asio/ssl.hpp>
+
+namespace fc {
+
+class http_client_impl {
+public:
+   using host_key = std::tuple<std::string, std::string, uint16_t>;
+   using raw_socket_ptr = std::unique_ptr<tcp::socket>;
+   using ssl_socket_ptr = std::unique_ptr<ssl::stream<tcp::socket>>;
+   using connection = static_variant<raw_socket_ptr, ssl_socket_ptr>;
+   using connection_map = std::map<host_key, connection>;
+   using error_code = boost::system::error_code;
+   using deadline_type = boost::posix_time::ptime;
+
+   http_client_impl()
+   :_ioc()
+   ,_sslc(ssl::context::sslv23_client)
+   {
+      set_verify_peers(true);
+   }
+
+   void add_cert(const std::string& cert_pem_string) {
+      error_code ec;
+      _sslc.add_certificate_authority(boost::asio::buffer(cert_pem_string.data(), cert_pem_string.size()), ec);
+      FC_ASSERT(!ec, "Failed to add cert: ${msg}", ("msg", ec.message()));
+   }
+
+   void set_verify_peers(bool enabled) {
+      if (enabled) {
+         _sslc.set_verify_mode(ssl::verify_peer);
+      } else {
+         _sslc.set_verify_mode(ssl::verify_none);
+      }
+   }
+
+   template<typename SyncReadStream, typename Fn, typename CancelFn>
+   error_code sync_do_with_deadline( SyncReadStream& s, deadline_type deadline, Fn f, CancelFn cf ) {
+      optional<error_code> timer_result;
+      boost::asio::deadline_timer timer(s.get_io_service());
+
+      timer.expires_at(deadline);
+      bool timer_cancelled = false;
+      timer.async_wait([&timer_result, &timer_cancelled] (const error_code& error) {
+         if (!error && !timer_cancelled) {
+            timer_result.emplace(error);
+         }
+      });
+
+      optional<error_code> f_result;
+      f(f_result);
+
+      s.get_io_service().reset();
+      while (s.get_io_service().run_one())
+      {
+         if (f_result) {
+            timer_cancelled = true;
+            timer.cancel();
+         } else if (timer_result) {
+            cf();
+         }
+      }
+
+      if (!timer_result) {
+         return *f_result;
+      } else {
+         return error_code(boost::system::errc::timed_out, boost::system::system_category());
```

Comment:
Might be worthwhile preserving the error_code as it is not necessarily timed_out.